src/utils/proxy.js

```diff
@@ -378,15 +367,36 @@ const startProxy = async function (settings, addonsUrls, configPath, projectDir)
     }
 
     proxy.web(req, res, options)
+  }
+
+  return onRequest
+}
+
+const createSecureServer = ({ onRequest, cert, key }) => https.createServer({ cert, key }, onRequest)
+
+const startProxy = async function (settings, addonsUrls, configPath, projectDir) {
+  const functionsServer = settings.functionsPort ? `http://localhost:${settings.functionsPort}` : null
+
+  const proxy = initializeProxy(settings.frameworkPort, settings.dist, projectDir)
+
+  const rewriter = await createRewriter({
+    distDir: settings.dist,
+    jwtSecret: settings.jwtSecret,
+    jwtRoleClaim: settings.jwtRolePath,
+    configPath,
+    projectDir,
   })
 
+  const onRequest = getRequestListener({ proxy, rewriter, settings, addonsUrls, functionsServer })
+  const server = settings.https ? createSecureServer({ ...settings.https, onRequest }) : http.createServer(onRequest)
   server.on('upgrade', function onUpgrade(req, socket, head) {
     proxy.ws(req, socket, head)
   })
 
   return new Promise((resolve) => {
     server.listen({ port: settings.port }, () => {
-      resolve(`http://localhost:${settings.port}`)
+      const scheme = settings.https ? 'https' : 'http'
+      resolve(`${scheme}://localhost:${settings.port}`)
```

Comment:
Can `settings.port` be `0` (i.e. "use any port")? If so, we might consider using `server.address().port` instead to get the resolved port.

Reply:
It cannot be `0`. We capture it before in:
https://github.com/netlify/cli/blob/41ec0d9a4626efa16efd0e2e292d84c66fd57d3d/src/utils/detect-server.js#L170

We do that to alert the user in case they configured a custom port, but we cannot capture it.